Resources/Catalog/RucioFileCatalogClient.py

```diff
@@ -0,0 +1,545 @@
+              for rse in rep['rses']:
+                pfn = rep['rses'][rse][0]
+                if self.convertUnicode:
+                  pfn = str(pfn)
+                result['Value']['Successful'][lfn]['Files'][name]['Replicas'][rse] = {'PFN': pfn, 'Status': 'U'}
+      except DataIdentifierNotFound as err:
+        result['Value']['Failed'][lfn] = 'No such file or directory'
+      except Exception as err:
+        return S_ERROR(str(err))
+    return result
+
+
+  @checkCatalogArguments
+  def getFileMetadata(self, lfns, ownership = False):
+    """ Returns the file metadata associated to a supplied LFN
+    """
+    successful, failed = {}, {}
+    lfnChunks = breakListIntoChunks(lfns, 1000)
+    listFiles = deepcopy(lfns.keys())
+    self.client = Client()
+    for chunk in lfnChunks:
+      try:
+        dids = [self.__getDidsFromLfn(lfn) for lfn in chunk]
+        for meta in self.client.get_metadata_bulk(dids):
+          lfn = str(meta['name'])
+          if meta['did_type'] in ['DATASET', 'CONTAINER']:
+            nlinks = len([child for child in self.client.list_content(meta['scope'], meta['name'])])
+            successful[lfn] = {'Checksum': '',
+                               'ChecksumType': '',
+                               'CreationDate': meta['created_at'],
+                               'GUID': '',
+                               'Mode': 509,
+                               'ModificationDate': meta['updated_at'],
+                               'NumberOfLinks': nlinks,
+                               'Size': 0,
+                               'Status': '-'}
+            try:
+              listFiles.remove(lfn)
+            except ValueError:
+              pass
+          else:
+            successful[lfn] = {'Checksum': str(meta['adler32']),
+                               'ChecksumType': 'AD',
+                               'CreationDate': meta['created_at'],
+                               'GUID': str(meta['guid']),
+                               'Mode': 436,
+                               'ModificationDate': meta['updated_at'],
+                               'NumberOfLinks': 1,
+                               'Size': meta['bytes'],
+                               'Status': '-'}
+            try:
+              listFiles.remove(lfn)
+            except ValueError:
+              pass
+      except DataIdentifierNotFound as err:
+        failed[lfn] = str(err)
+      except Exception as err:
+        return S_ERROR(str(err))
+    for lfn in listFiles:
+      failed[lfn] = 'No such file or directory'
+    resDict = {'Failed': failed, 'Successful': successful}
+    return S_OK(resDict)
+
+
+  @checkCatalogArguments
+  def exists(self, lfns):
+    """ Check if the path exists
+    """
+    result = {'OK': True, 'Value': {'Successful': {}, 'Failed': {}}}
+    self.client = Client()
```

Comment:
Do you really need to re-instantiate the `Client` object at every method call?

Reply:
Unfortunately yes. When one instantiates a `Client`, it creates a http session. This session will expire after some time, then it's not possible to do any new call.

Reply:
If you know the lifetime of the session, you can safely use the `DictCache` object

Reply:
If you can check if the connection is still alive you could use a property:
```python
  @property
  def client(self):
    if not self._client.is_alive:
      self._client = Client()
    return self._client
```

Reply:
Done